### DISCUSSION TITLE
Configuration Syntax

### DISCUSSION THREAD

**Discussion by nicholas-ochoa** [moudey/Shell#84] [Ideas]
This is probably a huge ask, but I keep running in to issues with the syntax (and it seems others do as well).

Is it possible to add support for (or switch entirely over to) a different syntax for configuration? Ideally, something that already exists.

JSON5 (or regular JSON), YAML, or TOML for example.

Something like JSON5 or YAML would allow for a number of huge benefits:

1. Support for a defined schema that defines allowed keys, properties, values, structure
2. Built in support for most text editors for syntax highlighting. If a schema is created some editors (VS Code, Sublime Text, others I'm sure) will show auto completion suggestions for configuration settings and values.

The current syntax is powerful it seems, but I feel like a lot of that power is locked up behind a large learning curve.

**Comment by moudey** [comment]
We'll definitely get to a standard format for the configuration file. In the future, a UI will be created to simplify the work as much as possible.

The current structure is dynamic and the option to import from other files has been added to split the settings file and make it more flexible. Take a look at this example

```
default
{
    import 'imports/settings.nss'
}

dynamic
{
    //...
    import 'imports/develop.nss'
    item(title='test')
    import if(sys.is11, 'imports/taskbar.nss', null)
    //...
}
```

**Reply by brain246** [reply]
Hi @moudey!

Instead of an own UI i would also prefer a standard file-format combined with a nss-config-schema, as proposed by @nicholas-ochoa. Because in the end, i am already editing all the files in VSCode, so any UI you create will be inferior to VSCode in terms of editing/syntax-highlighting/linting/formatting etc.. If the standard markup/structured formats aren't sufficient for your needs you can also create your own language. 

I am really no expert on this topic but i think VSCode/Sublime/Atom etc. can all work with https://github.com/microsoft/TypeScript-TmLanguage. 

And if i remember correctly there have been more features regarding languages... just throwing this link in ere as a teaser: https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

I think your valuable dev-time is better spent on other features than implementing a config UI. 

Other than that, i wanted to thank you a thousand times for creating this extremely useful tool! Without it Win 11 would be a nightmare to use. Since kind words are nice, but food is nicer, i sent you a little donation@paypal today... should be enough for a lunch at McDonalds 😆 Good appetite, thanks again and please keep working on it!

**Reply by moudey** [reply]
Thank you very much @brain246.
I am really working on making the configuration file context as standard as possible so that it is more stable. I will create a new repository for this feature so that we can provide an extension for most editors.

The UI is required for the normal user, because he has difficulty editing the config file which makes the file explorer unstable, so I am thinking of converting the config file into two layers.  The `NSS` layer is compiled after validation into an `NSO` binary file to be more stable to use.

I apologize for any unclear points due to the use of google translator.

**Comment by moudey** [comment]
The `import` feature can be extended using import from `json` or `xml` file.

**Comment by bugficks** [comment]
for the fun of it I did a quick hack "json to nss" [python script](https://gist.github.com/bugficks/50667b5a8be877d437a29c48b84d394d)
